File: 0967-numbers-with-same-consecutive-differences/0967-numbers-with-same-consecutive-differences.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[]}
 */
var numsSameConsecDiff = function (n, k) {
    const isValidNumber = function (arr) {
        let digit = arr[0];

        if (digit === 0) return false;

        for (let i = 1; i < arr.length; i++) {
            if (Math.abs(arr[i] - digit) !== k) return false;

            digit = arr[i];
        }

        return true;
    }

    const dfs = function (path) {
        if (path.length === n) {
            result.push(Number(path.join("")));
            return;
        }

        for (let i = 0; i <= 9; i++) {
            path.push(i);
            if (isValidNumber(path)) {
                dfs(path);
            }
            path.pop();
        }
    }

    const result = [];
    dfs([]);

    return result;

};

File: 0967-numbers-with-same-consecutive-differences/0967-numbers-with-same-consecutive-differences.py
class Solution:
    def numsSameConsecDiff(self, n: int, k: int) -> List[int]:
        result = []

        def dfs(path):
            if len(path) == n:
                number = int("".join(map(str, path)))
                result.append(number)
                return

            prev = path[-1]

            next_digit = prev + k

            if next_digit <= 9:
                path.append(next_digit)
                dfs(path)
                path.pop()

            next_digit = prev - k

            if next_digit >= 0 and next_digit != prev:
                path.append(next_digit)
                dfs(path)
                path.pop()

        for digit in range(1, 10):
            dfs([digit])

        return result
